mlc_llm/relax_model/llama.py

```diff
@@ -327,7 +318,121 @@ def forward(
             ),
         )
 
-        kv_seq_len = all_seq_len_shape.struct_info.values[0]
+        attn_output, past_key_values = self.attention_fwd(
+            query_states,
+            key_states,
+            value_states,
+            past_key_values,
+            bsz,
+            q_len,
+            layer_id=layer_id,
+            all_seq_len_shape=all_seq_len_shape,
+            attention_mask=attention_mask,
+        )
+
+        attn_output = nn.emit(
+            reshape(attn_output, (bsz, q_len, self.head_dim * self.num_query_heads))
+        )
+        attn_output = self.o_proj(attn_output)
+        return attn_output, past_key_values
+
+    def attention_fwd(
+        self,
+        query_states: relax.Expr,
+        key_states: relax.Expr,
+        value_states: relax.Expr,
+        past_key_values: relax.Expr,
+        batch_size: tir.PrimExpr,
+        q_len: tir.PrimExpr,
+        **kwargs,
+    ):
+        raise NotImplementedError()
+
+
+class LlamaPagedAttention(LlamaAttentionBase):
+    def __init__(self, config: LlamaConfig):
+        super().__init__(config)
+        ctx_mod = relax.BlockBuilder.current().get()
+        self.kv_cache_transpose_append = ctx_mod.get_global_var("kv_cache_transpose_append")
+        self.attention_compute = ctx_mod.get_global_var("attention")
+
+    def attention_fwd(
```

Comment:
This looks like decode function, should we also include prefill function in `LlamaPagedAttention` class?

Reply:
It is general for both prefill and decode.

Reply:
got it, that make sense to me